typegraph/node/sdk/src/tg_artifact_upload.ts

```diff
@@ -0,0 +1,142 @@
+// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
+// SPDX-License-Identifier: MPL-2.0
+
+import { BasicAuth } from "./tg_deploy.js";
+import { runtimes } from "./wit.js";
+import { Artifact } from "./gen/interfaces/metatype-typegraph-core.js";
+import { dirname, join } from "node:path";
+import * as fsp from "node:fs/promises";
+
+interface UploadArtifactMeta {
+  typegraphName: string;
+  relativePath: string;
+  sizeInBytes: number;
+  hash: string;
+}
+
+export class ArtifactUploader {
+  private getUploadUrl: URL;
+
+  constructor(
+    baseUrl: string,
+    private refArtifacts: Artifact[],
+    private tgName: string,
+    private auth: BasicAuth | undefined,
+    private headers: Headers,
+    private tgPath: string,
+  ) {
+    const suffix = `${tgName}/artifacts/upload-urls`;
+    this.getUploadUrl = new URL(suffix, baseUrl);
+  }
+
+  private async fetchUploadUrls(
+    artifactMetas: UploadArtifactMeta[],
+  ): Promise<Array<string | null>> {
+    const artifactsJson = JSON.stringify(artifactMetas);
+    const response = await fetch(this.getUploadUrl, {
+      method: "POST",
+      headers: this.headers,
+      body: artifactsJson,
+    });
+
+    if (!response.ok) {
+      const err = await response.text();
+      throw new Error(`Failed to get upload URLs for all artifacts: ${err}`);
+    }
+
+    const uploadUrls: Array<string | null> = await response.json();
+    if (uploadUrls.length !== artifactMetas.length) {
+      const diff =
+        `array length mismatch: ${uploadUrls.length} !== ${artifactMetas.length}`;
+      throw new Error(`Failed to get upload URLs for all artifacts: ${diff}`);
+    }
+
+    return uploadUrls;
+  }
+
+  private async upload(
+    url: string | null,
+    meta: UploadArtifactMeta,
+  ): Promise<any> {
+    const uploadHeaders = new Headers({
+      "Content-Type": "application/octet-stream",
+    });
+
+    if (this.auth) {
+      uploadHeaders.append("Authorization", this.auth.asHeaderValue());
+    }
+
+    if (url == null) {
+      console.log(`Skipping upload for artifact: ${meta.relativePath}`);
+      return;
+    }
+
+    const path = join(dirname(this.tgPath), meta.relativePath);
+    // TODO: stream
+    const content = await fsp.readFile(path);
+    const res = await fetch(url, {
+      method: "POST",
+      headers: uploadHeaders,
+      body: new Uint8Array(content),
+    } as RequestInit);
```

Comment:
I think we should expost the imported file ops functions to uniformize these (thinking of how the paths should be resolved).